spark-3.1-spanner-lib/src/main/java/com/google/cloud/spark/spanner/SpannerSpark.java

```diff
@@ -0,0 +1,229 @@
+            }
+
+            switch (fieldTypeName) {
+            case DataType."BOOL":
+                objects.add(spannerRow.getBoolean(i));
+            case "BYTES":
+                return DataTypes.BinaryType;
+            case "DATE":
+                objects.add(spannerRow.getDate(i));
+            case "FLOAT64":
+                objects.add(spannerRow.getBigDecimal(i));
+            case "INT64":
+                objects.add(spannerRow.getLong(i));
+            case "JSON":
+                return DataTypes.BinaryType;
+            case "NUMERIC":
+                objects.add(spannerRow.getBigDecimal(i));
+            case "STRING":
+                objects.add(spannerRow.getString(i));
+            case "TIMESTAMP":
+                objects.add(spannerRow.getTimestamp(i));
+            default: // "ARRAY", "STRUCT"
+                throw new Exception("unhandled type: " + fieldTypeName);
+            }
+        }
+
+        return RowFactory.create(objects.toArray(new Object[0]));
+    }
+
+    @Override
+    public String shortName() {
+        return "cloud-spanner";
+    }
+
+    @Override
+    public DataType getCatalystType(int sqlType, String typename, int size, MetadataBuilder mb) {
+        switch (typename) {
+            case "ARRAY":
+                return DataTypes.ArrayType;
+            case "BOOL":
+                return DataTypes.BooleanType;
+            case "BYTES":
+                return DataTypes.BinaryType;
+            case "DATE":
+                return DataTypes.DateType;
+            case "FLOAT64":
+                return DataTypes.DecimalType;
+            case "INT64":
+                return DataTypes.LongType;
+            case "JSON":
+                return DataTypes.BinaryType;
+            case "NUMERIC":
+                return DataTypes.DecimalType;
+            case "STRING":
+                return DataTypes.StringType;
+            case "STRUCT":
+                return DataTypes.StructType;
+            case "TIMESTAMP":
+                return DataTypes.TimestampType;
+            default:
+                return DataTypes.NullType;
+        }
+    }
+
+    @Override
+    public Transform[] inferPartitioning(CaseInsensitiveStringMap options) {
+        return null;
+    }
+
+    @Override
+    public StructType inferSchema(CaseInsensitiveStringMap options) {
+        return null;
+    }
+
+    @Override
+    public boolean supportsExternalMetadata() {
+        return false;
+    }
+
+    @Override
+    public Table getTable(StructType schema, Transform[] partitioning, java.util.Map<String, String> properties) {
+        // 1. Create the DatabaseClient with the provided options:
+        // "instanceId": <INSTANCE_ID>
+        // "projectId": <PROJECT_ID>
+        // "databaseId": <DATABASE_ID>
+        String spannerUri = String.format(
+                "cloudspanner:/projects/%s/instances/%s/databases/%s", 
```

Comment:
Probably we want to extract the usage of connectionOptions and Spanner operations into multiple class for easy testing.

Reply:
Indeed!